cmd/imagePushToRegistry.go

```diff
@@ -0,0 +1,235 @@
+	if err := handleCredentialsForPrivateRegistry(config.DockerConfigJSON, config.TargetRegistryURL, config.TargetRegistryUser, config.TargetRegistryPassword, utils); err != nil {
+		return errors.Wrap(err, "failed to handle credentials for target registry")
+	}
+
+	if len(config.LocalDockerImagePath) > 0 {
+		if err := pushLocalImageToTargetRegistry(config, utils); err != nil {
+			return errors.Wrapf(err, "failed to push local image to %q", config.TargetRegistryURL)
+		}
+		return nil
+	}
+
+	if err := copyImages(config, utils); err != nil {
+		return errors.Wrap(err, "failed to copy images")
+	}
+
+	return nil
+}
+
+func handleCredentialsForPrivateRegistry(dockerConfigJsonPath, registry, username, password string, utils imagePushToRegistryUtils) error {
+	if len(dockerConfigJsonPath) == 0 && (len(registry) == 0 || len(username) == 0 || len(password) == 0) {
+		return errors.New("docker credentials not provided")
+	}
+
+	if len(dockerConfigJsonPath) == 0 {
+		if _, err := docker.CreateDockerConfigJSON(registry, username, password, "", targetDockerConfigPath, utils); err != nil {
+			return errors.Wrap(err, "failed to create new docker config")
+		}
+		return nil
+	}
+
+	if _, err := docker.CreateDockerConfigJSON(registry, username, password, targetDockerConfigPath, dockerConfigJsonPath, utils); err != nil {
+		return errors.Wrapf(err, "failed to update docker config %q", dockerConfigJsonPath)
+	}
+
+	if err := docker.MergeDockerConfigJSON(targetDockerConfigPath, dockerConfigJsonPath, utils); err != nil {
+		return errors.Wrapf(err, "failed to merge docker config files")
+	}
+
+	return nil
+}
+
+func copyImages(config *imagePushToRegistryOptions, utils imagePushToRegistryUtils) error {
+	g, ctx := errgroup.WithContext(context.Background())
+	platform := config.TargetArchitecture
+
+	for i := 0; i < len(config.SourceImages); i++ {
+		src := fmt.Sprintf("%s/%s", config.SourceRegistryURL, config.SourceImages[i])
+		dst := fmt.Sprintf("%s/%s", config.TargetRegistryURL, config.TargetImages[i])
+
+		g.Go(func() error {
+			log.Entry().Infof("Copying %s to %s...", src, dst)
+			if err := utils.CopyImage(ctx, src, dst, platform); err != nil {
+				return err
+			}
+			log.Entry().Infof("Copying %s to %s... Done", src, dst)
+			return nil
+		})
+
+		if config.TagLatest {
+			g.Go(func() error {
+				// imageName is repository + image, e.g test.registry/testImage
+				imageName := parseDockerImageName(dst)
+				log.Entry().Infof("Copying %s to %s...", src, imageName)
+				if err := utils.CopyImage(ctx, src, imageName, platform); err != nil {
+					return err
+				}
+				log.Entry().Infof("Copying %s to %s... Done", src, imageName)
+				return nil
+			})
+		}
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func pushLocalImageToTargetRegistry(config *imagePushToRegistryOptions, utils imagePushToRegistryUtils) error {
+	g, ctx := errgroup.WithContext(context.Background())
```

Comment:
should make concurrency limit configurable?
we can use this pattern: https://go.dev/play/p/HovNRgp6FxH

I don't know if there any limits from Artifactory about max parallel uploads and so on.

Reply:
there is `SetLimit(n int)` from `golang.org/x/sync/errgroup`. It limits the number of active goroutines in this group to at most n. I set n to 10. Do you think it is ok? Also don't know if there are any limits from Artifactory side :)

Reply:
`SetLimit` is ok, sure :) I forgot about it